server/src/main/java/org/opensearch/cluster/routing/allocation/decider/NodeOverloadAwareAllocationDecider.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.cluster.routing.allocation.decider;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.opensearch.cluster.metadata.Metadata;
+import org.opensearch.cluster.routing.RoutingNode;
+import org.opensearch.cluster.routing.ShardRouting;
+import org.opensearch.cluster.routing.allocation.RoutingAllocation;
+import org.opensearch.common.settings.ClusterSettings;
+import org.opensearch.common.settings.Setting;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.common.settings.Setting.Property;
+
+import java.util.function.BiPredicate;
+
+/**
+ * This {@link NodeOverloadAwareAllocationDecider} controls shard over-allocation
+ * due to node failures or otherwise on the surviving nodes
+ * <pre>
+ * cluster.routing.allocation.overload_awareness.provisioned_capacity: N
+ * </pre>
+ * <p>
+ * and prevent allocation on the surviving nodes of the under capacity cluster
+ * based on oveload factor defined as a percentage by
+ * <pre>
+ * cluster.routing.allocation.overload_awareness.factor: X
+ * </pre>
+ */
+public class NodeOverloadAwareAllocationDecider extends AllocationDecider {
+
+    public static final String NAME = "overload_awareness";
+
+    public static final Setting<Integer> CLUSTER_ROUTING_ALLOCATION_OVERLOAD_AWARENESS_PROVISIONED_CAPACITY_SETTING =
+        Setting.intSetting("cluster.routing.allocation.overload_awareness.provisioned_capacity", -1, -1,
+            Property.Dynamic, Property.NodeScope);
+    public static final Setting<Integer> CLUSTER_ROUTING_ALLOCATION_OVERLOAD_AWARENESS_FACTOR_SETTING =
+        Setting.intSetting("cluster.routing.allocation.overload_awareness.factor", 50, -1, Property.Dynamic,
+            Property.NodeScope);
```

Comment:
Having this (feature/setting) at an index level makes more sense, given usually larger clusters have number of shards in the order of thousands (such as 1k per node), and even a small overload percentage lets say (10%) will allow overload of shards in the order of 100s per node. Key point here is not all shards overload is a problem, but only of selective indices, which are expected to have more resource intensive operations. 

Given this feature is supposed to restrict the resource utilization of most active indices, having this decider only to work for select indices, could make this feature more targeted and even customer friendly. Otherwise we may not be be able to effectively limit the overload of desired shards, in a large clusters. Thoughts?

Reply:
All allocation deciders use a property called priority to prioritize allocation of shards across indices. This decider inherently uses that same logic prioritizing more recent index/the one having high priority. Having a knob at index level might be hard to manage. Given there is a way to define priority today, we can see if that makes sense as a follow up

Reply:
>Given there is a way to define priority today, we can see if that makes sense as a follow up

That was the concern, since overload limit will be first utilised by the index with highest priority, which in general will also be the index with most resource consumption.

In the above example, lets say if the overload of 100 shards per node is allowed, the impacting index will just fill in those slot first, while less important indices may remain unassigned. This can defy the overall goal.

Agree this can be built incremental, however seems quite necessary. Hence the suggestion.

Reply:
The right solution is to factor in actual heat and get rid of these limits. Unfortunately that might need more time before we see heat based deciders. I would rather keep this simple for now

Reply:
> This can defy the overall goal 

This is not true, the shard allocation algorithm will balance out shards of the same index across nodes. This limit will prevent non-prioritized index to get overloaded on the same node as the one with high priority.

Reply:
>Unfortunately that might need more time before we see heat based deciders

Agree, however I still feel until then capability to avoid overload at the index level, will allow customers to have even distribution of indices that they most care about. 

> This limit will prevent non-prioritized index to get overloaded on the same node as the one with high priority beyond the overload limit

Exactly my point, since for customers the high-priority index would be those with more active reads and writes, they in-turn become the indices which are more resource intensive, and should be more evenly distributed. These indices hence should not be allowed to be overloaded beyond the set threshold by customer. But as stated in the above example, since the overload limit is applicable across all shards, the high-priority index will end up being allocated and skewed. 

I will suggest we evaluate this and do a quick follow-up.